src/runtime/pybuffer.cs

```diff
@@ -0,0 +1,167 @@
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace Python.Runtime
+{
+    public class PyBuffer
+    {
+        private IntPtr _handle;
+        private Runtime.Py_buffer _view;
+        private GCHandle _gchandle;
+
+        internal PyBuffer(IntPtr obj, int flags, out bool success)
+        {
+            int size = Marshal.SizeOf(typeof(Runtime.Py_buffer));
+            byte[] rawData = new byte[size];
+            _gchandle = GCHandle.Alloc(rawData, GCHandleType.Pinned);
+            _handle = _gchandle.AddrOfPinnedObject();
+            _view = (Runtime.Py_buffer)Marshal.PtrToStructure(_handle, typeof(Runtime.Py_buffer));
+
+            success = Runtime.PyObject_GetBuffer(obj, _handle, flags) >= 0;
+
+            if (_view.shape != IntPtr.Zero)
+            {
+                Shape = new long[_view.ndim];
+                Marshal.Copy(_view.shape, Shape, 0, (int)_view.len * sizeof(long));
+            }
+
+            if (_view.strides != IntPtr.Zero) {
+                Strides = new long[_view.ndim];
+                Marshal.Copy(_view.strides, Strides, 0, (int)_view.len * sizeof(long));
+            }
+
+            if (_view.suboffsets != IntPtr.Zero) {
+                SubOffsets = new long[_view.ndim];
+                Marshal.Copy(_view.suboffsets, SubOffsets, 0, (int)_view.len * sizeof(long));
+            }
+        }
+
+        public PyObject Object => new PyObject(_view.obj);
+
+        public long Length => _view.len;
+        public int Dimensions => _view.ndim;
+
+        /// <summary>
+        /// An array of length <see cref="Dimensions"/> indicating the shape of the memory as an n-dimensional array.
+        /// </summary>
+        public long[] Shape { get; }
+
+        /// <summary>
+        /// An array of length <see cref="Dimensions"/> giving the number of bytes to skip to get to a new element in each dimension.
+        /// Will be null except when PyBUF_STRIDES or PyBUF_INDIRECT flags in GetBuffer/>.
+        /// </summary>
+        public long[] Strides { get; }
+
+        /// <summary>
+        /// An array of Py_ssize_t of length ndim. If suboffsets[n] >= 0,
+        /// the values stored along the nth dimension are pointers and the suboffset value dictates how many bytes to add to each pointer after de-referencing.
+        /// A suboffset value that is negative indicates that no de-referencing should occur (striding in a contiguous memory block).
+        /// </summary>
+        public long[] SubOffsets { get; }
+
+        /// <summary>
+        /// Release the buffer view and decrement the reference count for view->obj. This function MUST be called when the buffer is no longer being used, otherwise reference leaks may occur.
+        /// It is an error to call this function on a buffer that was not obtained via <see cref="PyObject.GetBuffer()"/>.
+        /// </summary>
+        public void Release()
+        {
+            Runtime.PyBuffer_Release(_handle);
+            _gchandle.Free();
```

Comment:
It would be good to reset these handles to zero, and check, that the buffer has not been released yet.